woocommerce/woocommerce-ios
# Fix test cases in `CurrencyFormatterTests` from locale, singleton usage, and negative value with RTL

Hopefully fixes all cases in #1409

In [a previous PR](https://github.com/woocommerce/woocommerce-ios/pull/2008), it enabled DI'ing `CurrencySettings` to `CurrencyFormatter` for testing `PriceInputFormatter`, but we still use the initializer that uses the singleton values in `CurrencyFormatterTests`.

## Changes

There are multiple factors that cause the test failures in `CurrencyFormatterTests`:
- `CurrencyFormatter` was using `Locale.current` for language code for manual RTL/LTR & negative formatting, so it failed on RTL device languages
  - This PR DI'ed the `Locale` to each function that uses it, with default value to `.current`
- `CurrencyFormatter` and `CurrencyFormatterTests` were using `CurrencySettings.shared` (singleton) values, which could be any values if the simulator has launched previously that updated the `CurrencySettings.shared` values (e.g. having run the app on a site with different currency options than expected in the test cases)
  - This PR replaced all the `CurrencySettings.shared` singleton usage in `CurrencyFormatter` with its `currencySettings` instance, and replaced all the`CurrencySettings.shared` usage in `CurrencyFormatterTests` with its sample `currencySettings` instance from `settings-general.json` site settings that were expected in the test cases
- As @pmusolino pointed out in the thread in the project channel, the Left-to-Right character (0x200E in hex) is in the string from `Formatters.smallNumberFormatter` **for a negative value with RTL device languages**: https://github.com/woocommerce/woocommerce-ios/blob/3a72ae9fc621c6dda1c3a360484f746b057a75d2/WooCommerce/Classes/Extensions/Double+Woo.swift#L38
  - Since we already manually format a negative value by the negative sign, space, and absolute value, this PR just uses the absolute value of the amount for formatting: https://github.com/woocommerce/woocommerce-ios/commit/2c1fae198b5641b0ac1f56742df65555fa343ed0#diff-9602fa47bdc00e54d6c27795709e94acR185

## Testing

Prerequisite: have a site with currency options different from [`settings-general.json` (under ID `woocommerce_currency`)](https://github.com/woocommerce/woocommerce-ios/blob/3a72ae9fc621c6dda1c3a360484f746b057a75d2/Networking/NetworkingTests/Responses/settings-general.json)

- Launch the app, logged in to the site in the prerequisite
- Run tests in `WooCommerceTests` with a LTR device language that is not English --> all 632 tests should pass
- Repeat the above 2 steps, with a RTL device language (e.g. Arabic) --> all 632 tests should pass

Update release notes:
